pex/platforms.py

```diff
@@ -5,14 +5,82 @@
 
 from collections import namedtuple
 
-from pex.pep425tags import (
-    get_abbr_impl,
-    get_abi_tag,
-    get_impl_ver,
-    get_platform,
-    get_supported,
-    get_supported_for_any_abi
-)
+from pex.orderedset import OrderedSet
+from pex.pep425tags import get_abbr_impl, get_abi_tag, get_impl_ver, get_platform, get_supported
+
+
+def _gen_all_compatible_versions(version):
+  # We select major and minor here in the context of implementation version strings.
+  # These are typically two digit characters; eg "27" or "36", but they can be three digit
+  # characters in the case of pypy; eg "271". In the typical case the 1st digit represents the
+  # python major version and the 2nd digit it's minor version. In the pypy case, the 1st digit still
+  # represents the (hosting) python major version, the 2nd the pypy major version and the 3rd the
+  # pypy minor version. In both cases the last digit is the minor version and the python in question
+  # guarantees backwards compatibility of minor version bumps within the major version as per
+  # semver.
+  #
+  # Concrete examples of what we want to return in each case:
+  # 1. typical case of cpython "36": ["36", "35", "34", "33", "32", "31", "30"]
+  # 2. pypy case of "271": ["271", "270"].
+  #
+  # For more information on the pypy case see conversation here:
+  #   https://github.com/pypa/pip/issues/2882
+  # In particular https://github.com/pypa/pip/issues/2882#issuecomment-110925458 and
+  # https://github.com/pypa/pip/issues/2882#issuecomment-130404840.
+  # The fix work for pip handling of this is done here: https://github.com/pypa/pip/pull/3075
+
+  major, minor = version[:-1], version[-1]
+
+  def iter_compatible_versions():
+    # Support all previous minor Python versions.
+    for compatible_minor in range(int(minor), -1, -1):
+      yield '{major}{minor}'.format(major=major, minor=compatible_minor)
+
+  return list(iter_compatible_versions())
+
+
+def _get_supported(version=None, platform=None, impl=None, abi=None, force_manylinux=False):
+  versions = _gen_all_compatible_versions(version) if version is not None else None
+  all_supported = get_supported(
+    versions=versions,
+    platform=platform,
+    impl=impl,
+    abi=abi
+  )
+
+  def iter_all_supported():
+    for supported in all_supported:
+      yield supported
+      python_tag, abi_tag, platform_tag = supported
+      if platform_tag.startswith('linux') and force_manylinux:
+        yield python_tag, abi_tag, platform_tag.replace('linux', 'manylinux1')
+
+  return list(OrderedSet(iter_all_supported()))
+
+
+def _gen_all_abis(impl, version):
```

Comment:
Note Python 3 does not have `mu` anymore, thanks to flexible Unicode string storage https://www.python.org/dev/peps/pep-0393.

I think I get why this is okay to keep though. The idea here is to generate all possible ABI tags, then rely on `_get_supported` to filter out the ABI tags we don't need? If this is the case, it is not immediately obvious and a comment could help here.

Reply:
And, fwiw, this code was just moved from pep425tags.py when I updated that file. We had - imo wrongly - edited that file instead of just copying it. We now just copy it with the mknimsl noted modifications.